src/syntax.rs

```diff
@@ -0,0 +1,105 @@
+pub enum DiscreteType {
+    Int,
+    Bool,
+    Float,
+}
+
+pub enum Type {
+    Vector(usize, DiscreteType),
+    Function,
+    Shader,
+}
+
+pub enum ShaderType {
+    Vector,
+    Fragment,
+    Tesselation,
+    // And more!
+}
+
+pub enum BOp {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+    And,
+    Or,
+    Eq,
+    Neq,
+    Gt,
+    Gte,
+    Lt,
+    Lte,
+    BitAnd,
+    BitOr,
+    BitXor,
+}
+
+pub struct Block {
+    pub body: Vec<Box<Expr>>
+}
+pub struct Parameter {
+    name: String,
+    datatype: Type
+}
+
+pub enum AccessType {
+    Name(String),
+    Idx(usize),
+}
+
+pub enum Expr {
+    I(usize),
+    B(bool),
+    F(f32),
+    Vect {
+        is_matrix: bool,
+        datatype: DiscreteType,
+        value: Block,
+    },
+    Def {
+        mutable: bool,
```

Comment:
I think it might be a better idea to introduce separate `Def` and `MutDef` exprs.